### PR TITLE
Cherry pick: 0.15/logging for timestamps (#1096)

### PR DESCRIPTION
**Pull request by gsheasby**
_Question: are release notes needed?_
- Add trace logging to help track down the MultipleTimestamps issue

If customers hit issue #1000, they can turn trace logging on for the
following packages, in order to give us extra information the next time
they hit it:
- com.palantir.atlasdb.cassandra
- com.palantir.atlasdb.keyvalue.cassandra
- com.palantir.atlasdb.factory
- com.palantir.timestamp
- Further improvements to timestamp trace logging
- Print thread name alongside stack trace when getting a
  TimestampService
- Log more variable values in AvailableTimestamps
- Log host for Cassandra CAS calls, and print currentValues in error case.
- Thread dump when we create a timestamp service and when CAS fails
- Remove printHost flag from CassandraClientPool.runWithRetry

Instead, we log it whenever trace is enabled.
Also, use getHostString instead of getHostName, to avoid blocking on
DNS lookup.
- Miscellaneous trace-logging fixes
- Move ThreadDumps and StackTraceUtils to their own package

Also, wrap the calls with isTraceEnabled() for performance reasons.
- CTBS.getCurrentTimestampValues: readability fix
- Add banner warnings to copied classes

---

This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1098)

### CODE REVIEW

**Comment by SerialVelocity** [comment]
We should have a changelog for develop saying lots of trace logging has been added.

**Comment by gsheasby** [comment]
Feedback from a client that has these changes is that it produces an estimated 16GB of logs per day, we should consider cutting down.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java [hunk 1]
```diff
@@ -128,25 +136,38 @@ private void cas(Client client, Long oldVal, long newVal) {
                     ConsistencyLevel.SERIAL,
                     ConsistencyLevel.EACH_QUORUM);
         } catch (Exception e) {
+            log.error("[CAS] Error trying to set from {} to {}: {}", oldVal, newVal, e.getMessage());
```
Comment (tpetracca) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java]: I feel like we should just log the whole trace `e.getMessage()` is just the top message right?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java [hunk 2]
```diff
@@ -128,25 +136,38 @@ private void cas(Client client, Long oldVal, long newVal) {
...
                     + " leaders or a CLI being run with an embedded timestamp service against an already running"
                     + " service.";
             MultipleRunningTimestampServiceError err = new MultipleRunningTimestampServiceError(msg);
             log.error(msg, err);
+            log.error("Thread dump: " + ThreadDumps.programmaticThreadDump());
             lastWriteException = err;
             throw err;
         } else {
+            log.trace("[CAS] Setting cached limit to {}.", newVal);
             lastWriteException = null;
```
Comment (tpetracca) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java]: why does `lastWriteException` even exist?  It's only ever modified, not read.  If you remove the  `@GuardedBy("this")` in line 68/71 you'll see that the compiler will even warn you that it's never used.

Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java]: *shrug* probably left over from some pre-refactor stuff...

**Comment by gsheasby** [comment]
To unblock this, I intend to log the most high-signal stuff (both places where we thread-dump and maybe a LIMITED set of other one-time things) at INFO, so we get that by default, without exposing people to TRACE/DEBUG logging.

File: docs/source/release_notes/release-notes.rst [hunk 3]
```diff
@@ -46,6 +46,15 @@ v0.21.0
            This is useful for keeping batch sizes small if the rows are wide.
            (`Pull Request <https://github.com/palantir/atlasdb/pull/947>`__)
 
+    *    - |improved|
+         - Added a significant amount of logging at the TRACE level, aimed at tracking down the ``MultipleRunningTimestampServicesError``.
+           If clients are hitting this error, then they should add trace logging for the following packages:
+              - ``com.palantir.atlasdb.cassandra``
+              - ``com.palantir.atlasdb.factory``
+              - ``com.palantir.atlasdb.keyvalue.cassandra``
+              - ``com.palantir.timestamp``
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: Do we also want a note that some stuff is logged at INFO level (actually, just fetching the TimestampService)

**Comment by gsheasby** [comment]
Tested this out successfully. We should talk to @shashiadi about messaging, so people know to direct their logs to a good place.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 4]
```diff
@@ -104,6 +101,7 @@ public static SerializableTransactionManager create(
             Set<Schema> schemas,
             Environment env,
             boolean allowHiddenTableAccess) {
+        DebugLogger.logger.warn("Called TransactionManagers.create. This should only happen once.");
```
Comment (SerialVelocity) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: info?

File: docs/source/configuration/logging.rst [hunk 5]
```diff
@@ -15,3 +15,22 @@ the logging, simply enable ``TRACE`` logging for
...
+From version 0.22, it is recommended that you send logging related to the timestamp service to a separate appender.
+To do this, add the following to your logging configuration:
+
+.. code:: yaml
+
+    logging:
+      loggers:
+        com.palantir.timestamp:
+          level: TRACE
+          additive: false
```
Comment (SerialVelocity) [inline docs/source/configuration/logging.rst]: Do we want to overwrite the old log?

File: timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java [hunk 6]
```diff
@@ -28,6 +28,7 @@
     private final PersistentUpperLimit upperLimit;
 
     public AvailableTimestamps(LastReturnedTimestamp lastReturnedTimestamp, PersistentUpperLimit upperLimit) {
+        DebugLogger.logger.trace("Creating AvailableTimestamps object. This should only happen once.");
```
Comment (SerialVelocity) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: All the other ones seem to be info. Should this one be info too?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 4 above)
Comment (SerialVelocity) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: We should also mention what thread it's from, right? Just like `ServiceDiscoveringAtlasSupplier`

File: docs/source/configuration/logging.rst [hunk 7]
```diff
@@ -15,3 +15,22 @@ the logging, simply enable ``TRACE`` logging for
...
+===================================================
+
+From version 0.22, it is recommended that you send logging related to the timestamp service to a separate appender.
+To do this, add the following to your logging configuration:
+
+.. code:: yaml
+
+    logging:
+      loggers:
+        com.palantir.timestamp:
```
Comment (SerialVelocity) [inline docs/source/configuration/logging.rst]: We probably just want the `com.palantir.timestamp.DebugLogger` stuff to go to a different file, right?

File: .baseline/checkstyle/checkstyle-suppressions.xml [hunk 8]
```diff
@@ -22,6 +22,10 @@
     <suppress files="[/\\].*[/\\]generated_src[/\\]" checks="." />
     <suppress files="[/\\].*[/\\]generated_testSrc[/\\]" checks="." />
 
+    <!-- Copied PG classes -->
+    <suppress files="[/\\].*StackTraceUtils.java" checks="." />
+    <suppress files="[/\\].*ThreadDumps.java" checks="." />
```
Comment (gsheasby) [inline .baseline/checkstyle/checkstyle-suppressions.xml]: Done

**Comment by jboreiko** [comment]
So as per our discussion.  We have elected here to use a named logger (though associated with an empty utility class).  By default I would imagine most products to log ERROR / WARN for everything and then pick particular subpaths for additional logging e.g. com.palantir.atlasdb INFO or com.palantir INFO.  In this case this change would lead to super aggressive logging.  Instead we should force products / users to opt in by logging at reasonable levels for each call.  People interested in debuggin the issue can drop the logging for com.palantir.timestamp.DebugLogger to TRACE when needed.

**Comment by jboreiko** [comment]
Another option here would be to switch this to an actual name logger (not use a fully qualified class name as the name) and then log at info and users and just switch it on and off as they need.  There is still a chance that a product globally logs at INFO so you would want to move to TRACE given the density of this logging.

**Comment by jboreiko** [comment]
#1179 deals with my first comment

**Pull request merged**